core/unit_test/TestCTestDevice.cpp

```diff
@@ -0,0 +1,91 @@
+#include <gtest/gtest.h>
+
+namespace Kokkos {
+namespace Impl {
+
+int get_ctest_gpu(const char* local_rank_str);
+
+} // namespace Impl
+} // namespace Kokkos
+
+class ctest_environment : public ::testing::Test {
+protected:
+  void SetUp();
+};
+
+void ctest_environment::SetUp()
+{
+  setenv("CTEST_KOKKOS_DEVICE_TYPE", "gpus", 1);
+  setenv("CTEST_RESOURCE_GROUP_COUNT", "10", 1);
+  unsetenv("CTEST_RESOURCE_GROUP_0");
+  setenv("CTEST_RESOURCE_GROUP_1", "threads", 1);
+  setenv("CTEST_RESOURCE_GROUP_2", "threads,cores", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_3", "gpus", 1);
+  unsetenv("CTEST_RESOURCE_GROUP_3_GPUS");
+
+  setenv("CTEST_RESOURCE_GROUP_4", "gpus", 1);
+  setenv("CTEST_RESOURCE_GROUP_4_GPUS", "id:2", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_5", "gpus", 1);
+  setenv("CTEST_RESOURCE_GROUP_5_GPUS", "slots:1,id:2", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_6", "gpus", 1);
+  setenv("CTEST_RESOURCE_GROUP_6_GPUS", "id:2,slots:1", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_7", "threads,gpus", 1);
+  setenv("CTEST_RESOURCE_GROUP_7_GPUS", "id:3,slots:1", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_8", "gpus,threads", 1);
+  setenv("CTEST_RESOURCE_GROUP_8_GPUS", "id:1,slots:1", 1);
+
+  setenv("CTEST_RESOURCE_GROUP_9", "cores,gpus,threads", 1);
+  setenv("CTEST_RESOURCE_GROUP_9_GPUS", "id:4,slots:1", 1);
+}
+
+TEST_F(ctest_environment, no_device_type)
+{
+  unsetenv("CTEST_KOKKOS_DEVICE_TYPE");
+  EXPECT_EQ(Kokkos::Impl::get_ctest_gpu("0"), 0);
+}
+
+TEST_F(ctest_environment, no_process_count)
+{
+  unsetenv("CTEST_RESOURCE_GROUP_COUNT");
+  EXPECT_EQ(Kokkos::Impl::get_ctest_gpu("0"), 0);
+}
+
```

Comment:
Consider checking the error message is what you would expect (c.f. https://github.com/google/googletest/issues/952#issuecomment-521361666)